File: ascii_art/c++/ascii.cpp
#include "CImg.h"
#include <iostream>
#include <cstring>
#include <fstream>
#include <algorithm>

using namespace std;
using namespace cimg_library;
char asciiChar[] = "$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\\|()1{}[] -_+~<>i!lI;:,\"^`'. ";

char getChar(int r, int g, int b, int a);

int main(int argc, char *argv[]) {
  cimg_usage("Turn pic to ascii art");
  const char* output = cimg_option("-o","ascii.txt","OUTPUT");
  int width = cimg_option("--width",0,"WIDTH");
  int height = cimg_option("--height",0,"HEIGHT");
  const char* input = cimg_option("-i","ascii.png","INPUT");
  cimg_option("-h"," ","help");
  if (!output)
    output = "ascii.txt";
  cimg_help("usage: ascii.py [-h] [-o OUTPUT] [--width WIDTH] [--height HEIGHT] [-i INPUT]");
  CImg<unsigned char> image(input);
  if (width == 0 || height == 0) {
    width = image.width();
    height = image.height();
  } else {
    image.resize(width, height);
  }
  
  
  ofstream fout (output);
  cimg_forY(image, y) {
    cimg_forX(image, x) {
      int r = image(x, y , 0);
      int g = image(x, y , 1);
      int b = image(x, y , 2);
      int a = image(x, y , 3);
      fout << getChar(r, g, b, a);
    }
    fout << "\n";
  }
  fout.close();
  return 0;
}

char getChar(int r, int g, int b, int a = 256) {
  if (a == 0)
    return ' ';
  int length = strlen(asciiChar);
  // RGB to gray
  double gray = 0.2126 * r + 0.7152 * g + 0.0722 * b;

  double unit = 256.0 / length;
  return asciiChar[static_cast<int>(gray / unit)];
}

File: ascii_art/python/ascii.py
# -*- coding: utf-8 -*-
from PIL import Image
import argparse

parser = argparse.ArgumentParser()

parser.add_argument('file')
parser.add_argument('-o', '--output')
parser.add_argument('--width', type = int)
parser.add_argument('--height', type = int)

args = parser.parse_args()

IMG = args.file
WIDTH = args.width
HEIGHT = args.height
OUTPUT = args.output

asciiChar = list("`1234567890-=~!@#$%^&*()_+qwertyuiop[]\\asdfghjkl;'zxcvbnm,./QWERTYUIOP{}|\":LKJHGFDSAZXCVBNM<>? ")
#list("$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\|()1{}[] -_+~<>i!lI;:,\"^`'.")

def getChar(r, g, b, alpha = 256):
	if alpha == 0:
		return ' '
	
	length = len(asciiChar)
	# RGB to Gray
	gray = int(0.2126 * r + 0.7152 * g + 0.0722 * b)
	
	unit = 256.0 / length
	return asciiChar[int(gray/unit)]

if __name__ == '__main__':
	im = Image.open(IMG)
	if WIDTH is None or HEIGHT is None:
		WIDTH, HEIGHT = im.size
	im = im.resize((WIDTH, HEIGHT), Image.NEAREST)

	txt = ''

	for i in range(HEIGHT):
		for j in range(WIDTH):
			txt += getChar(*im.getpixel((j,i)))
		
		txt += '\n'
	print(txt)

	if OUTPUT:
		with open(OUTPUT, 'w') as f:
			f.write(txt)

	
